[API docs contribution guidelines] Add kibana local API docs workflow quickstart (#3335)
elastic/docs-content: contribute-docs/api-docs/kibana-api-docs-quickstart.md

---
navigation_title: Kibana API docs
---

# Contribute to Kibana API docs locally

## Overview

The Kibana API documentation is created by merging multiple OpenAPI documents. This is a step-by-step local development workflow. While CI runs these steps automatically on PR branches in the `kibana` repo, working locally enables you to validate, preview and debug before submitting your changes.

The workflow you follow depends on which set of APIs you're updating:

::::{tab-set}
:group: kibana-workflow

:::{tab-item} Code-generated
:sync: code-generated

The core Kibana APIs are automatically generated from TypeScript route definitions in the codebase. Edit the `.ts` files in your plugin, and CI will regenerate the OpenAPI files when you push your PR. You can also run the generation locally for validation.
:::

:::{tab-item} Manual YAML
:sync: manual

> Some teams, including Security and Observability, work with hand-edited YAML files in the `oas_docs` directory. Edit these files directly and skip the code-generation steps.

Reviewer: Sorry, I missed this last folder inaccuracy:
```suggestion
Some teams, including Security and Observability, work with hand-edited YAML files in their plugin and package directories. Edit these files directly and skip the code-generation steps. For the complete list of files, refer to the [merge scripts](https://github.com/elastic/kibana/tree/main/oas_docs/scripts). 
```

      },
    }),
  },
  // ...
})
```
::::
::::{dropdown} YAML examples

**For code-generated APIs**, reference a YAML file in your route definition:

```typescript
import path from 'node:path';

const oasOperationObject = () => path.join(__dirname, 'foo.examples.yaml');

.addVersion({
  version: '2023-10-31',
  options: {
    oasOperationObject,
  },
  validate: {
    request: {
      body: fooResource,
    },
    response: {
      200: {
        body: fooResourceResponse,
      },
    },
  },
})
```

**For Manual YAML**, add examples directly in your YAML files in the `oas_docs` directory.

**Example YAML structure:**

```yaml
requestBody:
  content:
    application/json:
      examples: # Use examples (plural), not example (deprecated)
        fooExample:
          summary: Foo example
          description: An example request of creating foo
          value:
            name: 'Cool foo!'
            enabled: true
responses:
  200:
    content:
      application/json:
        examples:
          exampleResponse:
            summary: Successful response
            value:
              id: '12345'
              name: 'Cool foo!'
              enabled: true
```
::::

::::::

> ::::::{step} (optional) Capture OAS snapshot

Reviewer: I think "OAS" isn't a widely-used acronym, so something like this might be clearer:
```suggestion
::::::{step} (optional) Capture code-generated output
```

:sync: code-generated

:::{note}
**This step is optional.** CI will automatically capture the snapshot when you push your `.ts` changes. Running this locally is useful for validating changes before pushing or debugging issues. See [`capture_oas_snapshot.sh`](https://github.com/elastic/kibana/blob/main/.buildkite/scripts/steps/checks/capture_oas_snapshot.sh) for the full list of paths captured in CI.
:::

This step captures the OpenAPI specification that {{kib}} generates at runtime from your route definitions. It spins up a local {{es}} and {{kib}} cluster with your code changes. This generates the following output files in the `oas_docs` directory:

- `bundle.json`
- `bundle.serverless.json`

**Prerequisites:**
- [Docker](https://docs.docker.com/get-docker/) must be running
- If you're an Elastician, ensure you're logged into Docker with your Elastic account

**Capture all API paths** (recommended):

```bash
node scripts/capture_oas_snapshot --update
```

**For faster iteration**, capture the specific paths you're working on:

```bash
node scripts/capture_oas_snapshot --update --include-path /api/your/specific/path
```
::::

::::{tab-item} Manual YAML
:sync: manual

> This step is not applicable for Manual YAML. Your YAML files in the `oas_docs` directory are used directly - no snapshot capture is needed. Skip to the next step.

Reviewer: Ditto the referenced files don't exist in the `oas_docs` repo:

```suggestion
This step is not applicable for Manual YAML. Your YAML files  are used directly. Skip to the next step.
```

::::

:::::
::::::

::::{step} Generate docs

> Run these commands to generate the merged OpenAPI documentation files:

Reviewer: Not mandatory, but this is more succinct:
```suggestion
Run these commands to merge the OpenAPI documentation files:
```

```bash
cd oas_docs
make api-docs
```

This generates the the following files:

- `oas_docs/output/kibana.yaml`

> - `oas_docs/output/kibana.info.yaml`

Reviewer: Wrong filename:
```suggestion
- `oas_docs/output/kibana.serverless.yaml`
```
